secret-memory/src/allocator.rs

```diff
@@ -0,0 +1,239 @@
+use allocator_api2::alloc::{AllocError, Allocator, Layout};
+use libc;
+use memsec;
+use std::fmt;
+use std::io::Error;
+use std::ptr::{NonNull, null_mut};
+
+/// A box backed by sodium_malloc
+pub type Box<T> = allocator_api2::boxed::Box<T, Alloc>;
+
+/// A vector backed by sodium_malloc
+pub type Vec<T> = allocator_api2::vec::Vec<T, Alloc>;
+
+/// Memory allocation using sodium_malloc/sodium_free
+#[derive(Clone)]
+pub struct Alloc {
+    inner: AllocInner
+}
+
+#[derive(Clone)]
+enum AllocInner {
+    SecretAlloc(SecretAlloc),
+    SodiumAlloc(SodiumAlloc),
+}
+
+impl Alloc {
+    pub fn new() -> Self {
+        Alloc {
+            inner: if get_support_memfd_secret() {
+                AllocInner::SecretAlloc(SecretAlloc::new())
+            } else {
+                AllocInner::SodiumAlloc(SodiumAlloc::default())
+            }
+        }
+    }
+}
+
+unsafe impl Allocator for Alloc {
+    fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        match &self.inner {
+            AllocInner::SecretAlloc(alloc) => alloc.do_allocate(layout),
+            AllocInner::SodiumAlloc(alloc) => alloc.do_allocate(layout),
+        }
+    }
+
+    unsafe fn deallocate(&self, ptr: NonNull<u8>, layout: Layout) {
+        match &self.inner {
+            AllocInner::SecretAlloc(alloc) => alloc.do_deallocate(ptr, layout),
+            AllocInner::SodiumAlloc(alloc) => alloc.do_deallocate(ptr, layout),
+        }
+    }
+}
+
+impl Default for Alloc {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl fmt::Debug for Alloc {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match &self.inner {
+            AllocInner::SecretAlloc(_) => fmt.write_str("<memfd_secret based Rust allocator>"),
+            AllocInner::SodiumAlloc(_) => fmt.write_str("<memsec based Rust allocator>")
+        }
+    }
+}
+
+#[cfg(not(unix))]
+fn get_support_memfd_secret() -> bool {
+    false
+}
+
+#[cfg(unix)]
+fn get_support_memfd_secret() -> bool {
+    static SUPPORT_MEMFD_SECRET: std::sync::OnceLock<bool> = std::sync::OnceLock::new();
+    *SUPPORT_MEMFD_SECRET.get_or_init(|| {
+        let fd  = unsafe {
+            libc::syscall(libc::SYS_memfd_secret, 0) as i32
+        };
+        if fd != -1 || if let Some(os_err) = Error::last_os_error().raw_os_error() { os_err != 38 } else { true } {
```

Comment:
This code is weird as hell; Error::last_os_error is the incorrect abstraction level (should use libc's builtin errno support).

errno is never being cleared, so this could fail spuriously based on the last invocation